service/src/java/org/apache/hive/service/cli/operation/OperationLogManager.java

```diff
@@ -0,0 +1,416 @@
+ * this will avoid the operation log being cleaned when session/operation is closed, refer to
+ * {@link HiveSessionImpl#close()}, so we can get the operation log for the optimization
+ * and investigating the problem of the operation handily for users or administrators.
+ * The tree under the log location looks like:
+ * - ${@link SessionManager#operationLogRootDir}_historic
+ *    - thriftPort__startTime
+ *       - sessionId
+ *          - queryId (the operation log file)
+ * <p>
+ * while the origin tree would like:
+ * - ${@link SessionManager#operationLogRootDir}
+ *    - sessionId
+ *       - queryId (the operation log file)
+ * <p>
+ * The lifecycle of the log is managed by a daemon called {@link OperationLogDirCleaner},
+ * it gets all query info stored in {@link QueryInfoCache}, searches for the query info that can not be reached on the webui,
+ * and removes the log. If the operation log session directory has no operation log under it and the session is dead,
+ * then the OperationLogDirCleaner will try to cleanup the session log directory.
+ */
+
+public class OperationLogManager {
+  private static final Logger LOG = LoggerFactory.getLogger(OperationLogManager.class);
+  private static final String HISTORIC_DIR_SUFFIX = "_historic";
+  private static String historicLogRootDir;
+  private static long maxBytesToFetch;
+
+  private final HiveConf hiveConf;
+  private final SessionManager sessionManager;
+  private final OperationManager operationManager;
+  private OperationLogDirCleaner cleaner;
+  private String historicParentLogDir;
+
+  public OperationLogManager(SessionManager sessionManager, HiveConf hiveConf) {
+    this.operationManager = sessionManager.getOperationManager();
+    this.hiveConf = hiveConf;
+    this.sessionManager = sessionManager;
+    if (HiveConf.getBoolVar(hiveConf, HiveConf.ConfVars.HIVE_SERVER2_HISTORIC_OPERATION_LOG_ENABLED)
+        && hiveConf.getBoolVar(HiveConf.ConfVars.HIVE_SERVER2_LOGGING_OPERATION_ENABLED)
+        && hiveConf.isWebUiQueryInfoCacheEnabled()) {
+      initHistoricOperationLogRootDir();
+      maxBytesToFetch = HiveConf.getSizeVar(hiveConf,
+          HiveConf.ConfVars.HIVE_SERVER2_HISTORIC_OPERATION_LOG_FETCH_MAXBYTES);
+      if (historicLogRootDir != null
+          && !HiveConf.getBoolVar(hiveConf, HiveConf.ConfVars.HIVE_IN_TEST)) {
+        cleaner = new OperationLogDirCleaner();
+        cleaner.start();
+      }
+    }
+  }
+
+  private void initHistoricOperationLogRootDir() {
+    String origLogLoc = hiveConf.getVar(HiveConf.ConfVars.HIVE_SERVER2_LOGGING_OPERATION_LOG_LOCATION);
+    File logLocation = new File(origLogLoc);
+    historicParentLogDir = logLocation.getAbsolutePath() + HISTORIC_DIR_SUFFIX;
+    int serverPort = hiveConf.getIntVar(HiveConf.ConfVars.HIVE_SERVER2_THRIFT_PORT);
+    String logRootDir = new StringBuilder(historicParentLogDir)
+        .append("/").append(serverPort)
+        .append("__").append(System.currentTimeMillis()).toString();
```

Comment:
Wouldn't be better some IP address/server name and port for readability and some UUID to ensure uniqueness?

Reply:
Since the operation log sits on the local machine, we already know the ip or server name when we try to scan the directories, the start time `System.currentTimeMillis()`  can take the same effect as and more meaningful than the UUID to ensure uniqueness of the same instance. So here I use the unique server port and the start time to construct the directory name.

Reply:
In our clusters we usually aggregate the log directories from multiple instances in case of an issue, and it is handy for the users if the name of the node is already in the file name or in the directory name of the logs, but we can always amend this later if needed.

Reply:
Make sense, I will add the node info to the directory. Thank you for the input.